### DISCUSSION TITLE
Installation failing - "make returned with non-zero (2) status"

### DISCUSSION THREAD

**Discussion by livialmg** [dtcenter/METplus#1377] [Installation] [answered]
Hi!

I've been trying to installl MET on ubuntu and the libraries seems to be successfully instaled, however MET was not. I followed other issues on the "discussion forum" and was able to fix a few problems as the flags related to python. I even attempt to install it on other machine, where everything worked fine. Nonetheless, I still couldn't successfully instal MET on my personal machine.

Any help would be great!

I am attaching the log files and scripts i've used.
[config.log](https://github.com/dtcenter/METplus/files/7944030/config.log)
[configure.log](https://github.com/dtcenter/METplus/files/7944031/configure.log)
[Uploading make.log…]()

**Comment by jprestop** [comment, marked as answer]
Hi @livialmg.  I see that you are having trouble compiling MET on your personal machine.  Thank you for taking a look at the other Discussions and for fixing a few problems with regard to Python.  I'm glad to hear that you were able to resolve some problems.  I'd be happy to take a look at your log files, but unfortunately, the file upload didn't seem to work.  There are links, but they do not lead to the files.  Could you please try again or if preferable, you could look at the instructions in [How to Send Us Data](https://github.com/dtcenter/METplus/discussions/954) and send them via ftp?

Also, can you please tell me if you are doing a manual installation or if you are using the compile_MET_all.sh script with the directions under Sample Script For Compiling External Libraries And MET on the [MET Download page](https://dtcenter.org/community-code/model-evaluation-tools-met/download)?  If you are not using the script, could you please send the config.log and the make.log files (configure.log is not needed)?  If are are using the script, could you please send the config.log, make.log, and the configuration file you're passing to the script?

NOTE: For the complete answer see details in comments below.

**Reply by jprestop** [reply]
Hi @livialmg, I see that you have the following set in your install_met_env.generic file:
```
#Root directory of your python install
export PYTHON_LOC="$(python3-config --prefix)"

#Directory of your python executable
export MET_PYTHON=${PYTHON_LOC}

#Python ldflags created using python3-config
export MET_PYTHON_LD="$(python3-config --ldflags)"

#Python cflags created using python3-config
export MET_PYTHON_CC="$(python3-config --cflags)"
```

Unfortunately, sometimes the output of running "python3-config --ldflags" doesn't give us the value we need.  In this case, that seems to be what is happening.  I see in config.log that the output of running "python3-config --ldflags" is:

```
-L/home/livia/miniconda3/lib/python3.8/config-3.8-x86_64-linux-gnu -L/home/livia/miniconda3/lib  -lcrypt -lpthread -ld\
l  -lutil -lrt -lm -lm 
```

That is close, but it is missing the necessary "-lpython3.8" **OR** "-lpython3.8m".  I'm not sure exactly where it would be located, likely either /home/livia/miniconda3/lib/python3.8/config-3.8-x86_64-linux-gnu or /home/livia/miniconda3/lib.   You could run the following to find out:

```
find /home/livia/miniconda3 -name "python3.8*"
```

After determining if it is python3.8 or python3.8m (we'll assume python3.8 for this example, you would need to modify the MET_PYTHON_LD in your install_met_env.generic file as so:

```
#Python ldflags created using python3-config doesn't provide everything needed; Setting manually
export MET_PYTHON_LD="-L/home/livia/miniconda3/lib/python3.8/config-3.8-x86_64-linux-gnu -L/home/livia/miniconda3/lib -lpython3.8 -lcrypt -lpthread -ldl  -lutil -lrt -lm -lm "
```

Notice the addition of the -lpython3.8 and the manually setting of MET_PYTHON_LD.

Separately, running "python3-config --cflags" give us many additional things that are not needed:

```
-I/home/livia/miniconda3/include/python3.8 -I/home/livia/miniconda3/include/python3.8  -Wno-unused-result -Wsign-compa\
re -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -ffunction-sections -pipe -isystem /home/livia/miniconda\
3/include -fdebug-prefix-map=/tmp/build/80754af9/python_1599203911753/work=/usr/local/src/conda/python-3.8.5 -fdebug-prefix-map=/home/livia/miniconda\
3=/usr/local/src/conda-prefix -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -flto -DNDEBUG -fwrapv -O3 -Wall
```

Sometimes that can also cause problems.  We can reduce that value greatly, so I would also recommended changing MET_PYTHON_CC in your install_met_env.generic file as so:

```
#Python cflags created using python3-config gives too much information; Setting manually 
export MET_PYTHON_CC="-I/home/livia/miniconda3/include/python3.8"
```

Please try changing those values and then try rerunning the script.  Please let us know how it goes.

**Reply by jprestop** [reply]
Hi @livialmg.  I just wanted to follow up and see if you were able to change those values and if so, did that produce a successful compilation of MET?

**Reply by livialmg** [reply]
Hi @jprestop, thank you very much for your insights! I've changed the MET_PYTHON_CC and MET_PYTHON_LD in the  install_met_env.generic which unfortunately didn't solve the issue. After a careful look in the log files, a found an error that traced back to the python compilation. So a updated the python version, which solved the problem.

**Reply by jprestop** [reply]
Great!  Thank you so much for letting us know. At this point I'm going to lock this conversation to prevent future posts. Our team has decided that once discussions have been answered, and the answer has been confirmed, we'll lock them. We want to encourage users to ask new questions in new discussions rather than posting to old ones. Hopefully that'll make the questions/answers easier for other users to follow.  So if/when more issues/questions arise, please feel free to start a new discussion.
